client.go

```diff
@@ -1,328 +1,166 @@
 package meilisearch
 
 import (
-	"context"
-	"github.com/pkg/errors"
-	"github.com/valyala/fasthttp"
-	"net/url"
-	"time"
+	"net/http"
 
-	"encoding/json"
+	"github.com/valyala/fasthttp"
 )
 
-// Config configure the Client
-type Config struct {
+// ClientConfig configure the Client
+type ClientConfig struct {
 
 	// Host is the host of your meilisearch database
 	// Example: 'http://localhost:7700'
 	Host string
 
-	// APIKey is optional
-	APIKey string
+	// MasterKey is optional
+	MasterKey string
 }
 
 // ClientInterface is interface for all Meilisearch client
 type ClientInterface interface {
-	WaitForPendingUpdate(ctx context.Context, interval time.Duration, indexID string, updateID *AsyncUpdateID) (UpdateStatus, error)
-	DefaultWaitForPendingUpdate(indexUID string, updateID *AsyncUpdateID) (UpdateStatus, error)
-
-	Indexes() APIIndexes
-	Version() APIVersion
-	Documents(indexID string) APIDocuments
-	Search(indexID string) APISearch
-	Updates(indexID string) APIUpdates
-	Settings(indexID string) APISettings
-	Keys() APIKeys
-	Stats() APIStats
-	Health() APIHealth
-	Dumps() APIDumps
-}
-
-// Client is a structure that give you the power for interacting with an high-level api with meilisearch.
-type Client struct {
-	config     Config
-	httpClient *fasthttp.Client
-
-	// singleton clients which don't need index id
-	apiIndexes APIIndexes
-	apiKeys    APIKeys
-	apiStats   APIStats
-	apiHealth  APIHealth
-	apiVersion APIVersion
-	apiDumps   APIDumps
-}
-
-// Indexes return an APIIndexes client.
-func (c *Client) Indexes() APIIndexes {
-	return c.apiIndexes
-}
-
-// Version return an APIVersion client.
-func (c *Client) Version() APIVersion {
-	return c.apiVersion
-}
-
-// Documents return an APIDocuments client.
-func (c *Client) Documents(indexID string) APIDocuments {
-	return newClientDocuments(c, indexID)
-}
-
-// Search return an APISearch client.
-func (c *Client) Search(indexID string) APISearch {
-	return newClientSearch(c, indexID)
-}
-
-// Updates return an APIUpdates client.
-func (c *Client) Updates(indexID string) APIUpdates {
-	return newClientUpdates(c, indexID)
-}
-
-// Settings return an APISettings client.
-func (c *Client) Settings(indexID string) APISettings {
-	return newClientSettings(c, indexID)
-}
-
-// Keys return an APIKeys client.
-func (c *Client) Keys() APIKeys {
-	return c.apiKeys
-}
-
-// Stats return an APIStats client.
-func (c *Client) Stats() APIStats {
-	return c.apiStats
-}
-
-// Health return an APIHealth client.
-func (c *Client) Health() APIHealth {
-	return c.apiHealth
-}
-
-// Dumps return an APIDumps client.
-func (c *Client) Dumps() APIDumps {
-	return c.apiDumps
-}
+	Index(uid string) *Index
+	GetIndex(indexID string) (resp *Index, err error)
+	GetAllIndexes() (resp []Index, err error)
+	CreateIndex(config IndexConfig) (resp *Index, err error)
+	GetOrCreateIndex(config IndexConfig) (resp *Index, err error)
```

Comment:
In `CreateIndex` and `GetOrCreateIndex`, the parameter `IndexConfig` could be also be passed as a reference instead of by value

Reply:
@eskombro,
I used to send the value instead of reference if the variable is expected to not be modified and it's not a big structure. Also I preferred this syntax:
``` go
	index, err := client.GetOrCreateIndex(meilisearch.IndexConfig{
		Uid: "indexUID",
	})
```
I can change it but it's used in the instantiation of the client too:
``` go
	var client = meilisearch.NewClient(meilisearch.ClientConfig{
		Host:      "http://127.0.0.1:7700",
		MasterKey: "masterKey",
	})
```
I can change and send references in every function if you think it's better.

Reply:
Done